File: x64/Debug/js/encryption.js
// preloader start
function go_preloader()
{
    document.getElementById("close-popup").style.visibility = "hidden";
}

// Calculate rounder keys
function arrayKeys(inKey, amountRounds)
{
    rightAraayKeys = [];
    tmpRounderKeys = [];
	readyRounderRey = [];

    // Number of key expansion
	const sincerelyNumber = 2;
    
	sizeInKey = inKey.length;
	halfSizeInKey = sizeInKey / 2;

    // Take the right half of the key
	for (var i = halfSizeInKey; i < sizeInKey; ++i)
	{
		rightAraayKeys.push(inKey[i]);
	}

    // Calculate the key for each round
	var rounderKey;

	var sizeRightAraayKeys = rightAraayKeys.length;

	for (var i = 1; i <= amountRounds; ++i)
	{
		rounderKey = Math.pow(sincerelyNumber, i);
	
		tmpRounderKeys = rounderKey.toString(2);
		
		sizeTmpRounderKey = tmpRounderKeys.length;
		amountBitsToNull = halfSizeInKey - sizeTmpRounderKey;
		sizeReadyKey = sizeTmpRounderKey + amountBitsToNull;
		
		for (var k = 0; k < amountBitsToNull; ++k)
		{
			readyRounderRey.push(0);
		}

		for (var z = 0; z < sizeTmpRounderKey; ++z)
		{
			readyRounderRey.push(tmpRounderKeys[z]);
		}

		var tmp_sizeRightAraayKeys = rightAraayKeys.length;
		var lastRecordedKeyBegin = tmp_sizeRightAraayKeys - sizeReadyKey;

		for (var g = 0; g < sizeReadyKey; ++g)
		{
			readyRounderRey[g] = readyRounderRey[g] ^ rightAraayKeys[lastRecordedKeyBegin];

			++lastRecordedKeyBegin;
		}

		for (r = 0; r < sizeReadyKey; ++r)
		{
			rightAraayKeys.push(readyRounderRey[r]);
		}

		readyRounderRey = [];
	}

    return rightAraayKeys;
}


// Create subblock
function subblock(inBlock, leftBlock, rightBlock)
{
	var sizeInBlock = inBlock.length;
	var halfSizeInBlock = sizeInBlock / 2;

	// Add values in left block 
	for (var i = 0; i < halfSizeInBlock; ++i)
	{
		leftBlock.push(inBlock[i]);
	}

	// Add values in right block
	for (var i = halfSizeInBlock; i < sizeInBlock; ++i)
	{
		rightBlock.push(inBlock[i]);
	}
}


// Func F (black box)
function F(subblock, rightAraayKeys, numberRounds)
{
	var res = [];

	sizeSubblock = subblock.length;
	var endKey;
	var beginKey;

	if (numberRounds == 0)
	{
		endKey = sizeSubblock;
		beginKey = 0;
	}
	else
	{
		endKey = sizeSubblock * (numberRounds + 1);
		beginKey = endKey - sizeSubblock;
	}

	var i = 0;

	for (beginKey; beginKey < endKey; ++beginKey)
	{
		res.push(subblock[i] ^ rightAraayKeys[beginKey]);

		++i;
	}

	return res;
}


// Encryption
function encryptFeistelNetwork(leftBlock, rightBlock, rightAraayKeys, amountRounds)
{
	var res = [];
	var temp = [];
	var tempResF = [];

	var sizeRightBlock = rightBlock.length;

	for (var i = 0; i < amountRounds; ++i)
	{
		tempResF = F(leftBlock, rightAraayKeys, i);

		for (var k = 0; k < sizeRightBlock; ++k)
		{
			temp.push(rightBlock[k] ^ tempResF[k]);
		}

		rightBlock = leftBlock;

		leftBlock = temp;

		temp = [];
	}

	for (var i = 0; i < sizeRightBlock; ++i)
	{
		res.push(leftBlock[i]);
	}

	for (var i = 0; i < sizeRightBlock; ++i)
	{
		res.push(rightBlock[i]);
	}

	return res.join('');
}


//
function decryptFeistelNetwork(inBlock, amountRounds, rightAraayKeys)
{
	var res = [];
	var leftBlock = [];
	var rightBlock = [];
	var temp = [];
	var tempResF = [];

	subblock(inBlock, leftBlock, rightBlock);

	var sizeRightBlock = rightBlock.length;

	for (var i = amountRounds - 1; i >= 0; --i)
	{
		tempResF = F(rightBlock, rightAraayKeys, i);

		for (var k = 0; k < sizeRightBlock; ++k)
		{
			temp.push(leftBlock[k] ^ tempResF[k]);
		}

		leftBlock = rightBlock;

		rightBlock = temp;

		temp = [];
	}

	for (var i = 0; i < sizeRightBlock; ++i)
	{
		res.push(leftBlock[i]);
	}

	for (var i = 0; i < sizeRightBlock; ++i)
	{
		res.push(rightBlock[i]);
	}

	return res.join('');
}


// Encoding in binary
function inBinary(inData)
{
    var tmp_inData = String(inData);
    var lengthInData = tmp_inData.length;
    var tmp_Binary;
    var res;
    var additionZero = [];

    var someSymbol = [];

    for(var i = 0; i < lengthInData; ++i)
    {
        tmp_Binary = tmp_inData[i].charCodeAt().toString(2);
        
        var lengthTmpBinary = tmp_Binary.length;
        var beginBorder = 24 - lengthTmpBinary;

        for(var k = 0; k < 24; ++k)
        {
          
            additionZero[k] = 0;
        }

        var l = 0;

        for(var t = beginBorder; t < 24; ++t)
        {
            additionZero[t] = tmp_Binary[l];

            ++l
        }

        someSymbol[i] = additionZero.join('');
    }

    return res = someSymbol.join('');
}


// RC4
function rc4(key, str) {
	var s = [], j = 0, x, res = '';
	for (var i = 0; i < 256; i++) {
		s[i] = i;
	}
	for (i = 0; i < 256; i++) {
		j = (j + s[i] + key.charCodeAt(i % key.length)) % 256;
		x = s[i];
		s[i] = s[j];
		s[j] = x;
	}
	i = 0;
	j = 0;
	for (var y = 0; y < str.length; y++) {
		i = (i + 1) % 256;
		j = (j + s[i]) % 256;
		x = s[i];
		s[i] = s[j];
		s[j] = x;
		res += String.fromCharCode(str.charCodeAt(y) ^ s[(s[i] + s[j]) % 256]);
	}
  
	return res;
}


// DH and Encrypt Auth data
function encryptAuthData()
{
    // gen  
    var g_beg = 2;
    var g_end = 15;

    var g_val = Math.random()*(g_end - g_beg + 1);
    var gen = Math.floor(g_val) + g_beg;

    // simple mod
    var sm_beg = 2;
    var sm_end = 100;
    
    var sm_val = Math.random()*(sm_end - sm_beg + 1);
    var simple_mod = Math.floor(sm_val) + sm_beg;

    // private_number
    var p_beg = 1;
    var p_end = 6;
    
    var p_val = Math.random() / 32767 *(p_end - p_beg + 1);
    var private_number = Math.floor(p_val) + p_beg;

    var res_degree = Math.pow(gen, private_number);
    var res_module = res_degree % simple_mod;

    // send gen, mod to server
    var msg = "DH_proto" + gen + "&" + simple_mod + "&" + res_module;
        $.ajax({
          async: false,
          type: 'POST',
          url: '127.0.0.1',
          data: msg,
          success: function(data) {

            // generate secret key
            var degree = Math.pow(data, private_number);
            var secret_key = String(degree % simple_mod);
            //alert(secret_key);

            var amoundRounds = 16;

            var leftBlock = [];
            var rightBlock = [];
            var arrayRounderKeys = [];
            var encryptLogin;
            var encryptPass;

            // get auth data
            var get_login = String($("input[name='AuthLogin']").val());
            var get_pass = String($("input[name='AuthPass']").val());

            // hashing auth data (SHA512)
            secret_key = hex_sha512(secret_key);
            get_login = hex_sha512(get_login);
            get_pass = hex_sha512(get_pass);
            
            if(get_login.length != 128 || get_pass.length != 128 || secret_key.length != 128)
            {
              alert("Error SHA512, length != 128 symbols or secret key != 128");
            }

            // in Binary auth 
            secret_key = inBinary(secret_key);
            get_login = inBinary(get_login);
            get_pass = inBinary(get_pass);

            //
            subblock(get_login.split(''), leftBlock, rightBlock);

            //
            arrayRounderKeys = arrayKeys(secret_key.split(''), amoundRounds);

            //
            encryptLogin = encryptFeistelNetwork(leftBlock, rightBlock, arrayRounderKeys, amoundRounds);

            //
            leftBlock = [];
            rightBlock = [];

            //
            subblock(get_pass.split(''), leftBlock, rightBlock);

            //
            encryptPass = encryptFeistelNetwork(leftBlock, rightBlock, arrayRounderKeys, amoundRounds);

            // start preloader
            go_preloader();

            // data overwrite
            $('#user_name').val(encryptLogin);
            $('#user_pass').val(encryptPass);

          },
          error:  function(xhr, str){
	        alert('Возникла ошибка: ' + xhr.responseCode);
          }
        });  
}


// DH and encrypt Reg data
function encryptRegData()
{
	// gen  
    var g_beg = 2;
    var g_end = 15;

    var g_val = Math.random()*(g_end - g_beg + 1);
    var gen = Math.floor(g_val) + g_beg;

    // simple mod
    var sm_beg = 2;
    var sm_end = 100;
    
    var sm_val = Math.random()*(sm_end - sm_beg + 1);
    var simple_mod = Math.floor(sm_val) + sm_beg;

    // private_number
    var p_beg = 1;
    var p_end = 6;
    
    var p_val = Math.random() / 32767 *(p_end - p_beg + 1);
    var private_number = Math.floor(p_val) + p_beg;

    var res_degree = Math.pow(gen, private_number);
    var res_module = res_degree % simple_mod;

    // send gen, mod to server
    var msg = "DH_proto" + gen + "&" + simple_mod + "&" + res_module;
        $.ajax({
          async: false,
          type: 'POST',
          url: '127.0.0.1',
          data: msg,
          success: function(data) {

            // generate secret key
            var degree = Math.pow(data, private_number);
            var secret_key = String(degree % simple_mod);
            //alert(secret_key);

            var amoundRounds = 16;

            var leftBlock = [];
            var rightBlock = [];
            var arrayRounderKeys = [];

			var encryptName;
            var encryptLogin;
            var encryptPass;

            // get auth data
			var get_name = String($("input[name='RegName']").val());
            var get_login = String($("input[name='RegLogin']").val());
            var get_pass = String($("input[name='RegPass']").val());

            // hashing key
            secret_key = hex_sha512(secret_key);

			// hashing auth data (SHA512)
			get_name = hex_sha512(get_name);
            get_login = hex_sha512(get_login);
            get_pass = hex_sha512(get_pass);
            
            if(get_login.length != 128 || get_pass.length != 128 || secret_key.length != 128 || get_name.length != 128)
            {
              alert("Error SHA512, length != 128 symbols or secret key != 128");
            }

            // in Binary
            secret_key = inBinary(secret_key);

			get_name = inBinary(get_name);
            get_login = inBinary(get_login);
            get_pass = inBinary(get_pass);

			 // Generate round keys
            arrayRounderKeys = arrayKeys(secret_key.split(''), amoundRounds);

            // Divide into sub-blocks name
            subblock(get_name.split(''), leftBlock, rightBlock);

			// Encrypt name
			encryptName = encryptFeistelNetwork(leftBlock, rightBlock, arrayRounderKeys, amoundRounds);

			// Cleaning temp array
            leftBlock = [];
            rightBlock = [];

			 // Divide into sub-blocks login
            subblock(get_login.split(''), leftBlock, rightBlock);

            // Encrypt login
            encryptLogin = encryptFeistelNetwork(leftBlock, rightBlock, arrayRounderKeys, amoundRounds);

            // Cleaning temp array
            leftBlock = [];
            rightBlock = [];

            // Divide into sub-blocks pass
            subblock(get_pass.split(''), leftBlock, rightBlock);

            // Encrypt pass
            encryptPass = encryptFeistelNetwork(leftBlock, rightBlock, arrayRounderKeys, amoundRounds);

            // start preloader
            go_preloader();

            // data overwrite
            $('#reg_user_name').val(encryptName);
            $('#reg_user_login').val(encryptLogin);
			$('#reg_user_pass').val(encryptPass);

          },
          error:  function(xhr, str){
	        alert('Возникла ошибка: ' + xhr.responseCode);
          }
        });  
}

File: ws_server/Encryption.cpp
/*******************************************************************************/
#include "Encryption.h"


/*******************************************************************************/
Encryption Enc;


/*******************************************************************************/
int Encryption::get_mod_and_gen(int gen, int mod, int& private_number)
{
	// initialization random number generator
	srand((unsigned int)time(0));

	// generate private number
	private_number = rand() % (6 - 1) + 1;

	// calculate the degree
	int degree = (int)pow(gen, private_number);

	// calculate modulus
	int out_mod_res = degree % mod;

	// give results
	return out_mod_res;
}


/*******************************************************************************/
int Encryption::calculated_mod(int in_mod_res, int& mod, int& private_number)
{
	// calculate the degree (an input module)
	int degree = (int)pow(in_mod_res, private_number);

	// calculate the secret key
	int secret_key = degree % mod;

	// give results
	return secret_key;
}


/*******************************************************************************/
vector<int> Encryption::hashingKey(int key)
{
	vector<int> resKey(3072);

	string secretKeyToStr = to_string(key);

	unsigned char digest[SHA512_DIGEST_LENGTH];

	char secretKeyHash[SHA512_DIGEST_LENGTH * 2 + 1];
	char secretKeyToChar[5];

	strncpy_s(secretKeyToChar, secretKeyToStr.c_str(), sizeof(secretKeyToChar));

	SHA512((unsigned char*)&secretKeyToChar, strlen(secretKeyToChar), (unsigned char*)&digest);

	for (int i = 0; i < SHA512_DIGEST_LENGTH; i++)
	{
		sprintf_s(&secretKeyHash[i * 2], (SHA512_DIGEST_LENGTH * 2 + 1), "%02x", (unsigned int)digest[i]);
	}
	
	// Copy in string
	string secretKeyStrHash(secretKeyHash);

	resKey = strSymbolsToIntCodeSymbol(secretKeyStrHash);

	return resKey;
}


/*******************************************************************************/
string Encryption::vectorCharToString(const vector<int>& inData)
{
	vector<char> tmp_res;

	tmp_res = IntbinaryToChar(inData);

	string res(tmp_res.begin(), tmp_res.end());

	return res;
}


/*******************************************************************************/
string Encryption::decode_rc4(const string &data, string key)
{
	RC4_KEY key_rc;

	size_t len = data.size();

	size_t keyLength = key.length();

	unsigned char *obuf = (unsigned char*)malloc(len + 1);
	memset(obuf, 0, len + 1);

	RC4_set_key(&key_rc, (int)keyLength, (const unsigned char*)key.c_str());
	RC4(&key_rc, len, (const unsigned char*)data.c_str(), obuf);

	string decode_data((char*)obuf, len);
	free(obuf);

	return decode_data;
}


/*******************************************************************************/
string Encryption::encode_rc4(const string &data, string key)
{
	RC4_KEY key_rc;

	size_t len = data.size();

	size_t keyLength = key.length();

	unsigned char *obuf = (unsigned char*)malloc(len + 1);
	memset(obuf, 0, len + 1);

	RC4_set_key(&key_rc, (int)keyLength, (const unsigned char*)key.c_str());
	RC4(&key_rc, len, (const unsigned char*)data.c_str(), obuf);

	string encode_data((char*)obuf, len);
	free(obuf);

	return encode_data;
}


/*******************************************************************************/
vector<int> Encryption::inBinary(int inNumber)
{
	vector<int> res;

	int tmp_number = int();

	do
	{
		tmp_number = (inNumber % 2);
		inNumber /= 2;

		res.push_back(tmp_number);
	} while (inNumber != 0);

	reverse(res.begin(), res.end());

	return res;
}


/*******************************************************************************/
vector<int> Encryption::strSymbolsToIntCodeSymbol(string inString)
{
	size_t sizeInString = inString.size();

	vector<int> intCodeSymbols;
	vector<int> binaryCodeSymbols;
	vector<int> tmp_binaryCode;

	for (int i = 0; i < sizeInString; ++i)
	{
		intCodeSymbols.push_back(inString[i]);
	}

	size_t sizeIntCodeSymbols = intCodeSymbols.size();

	for (int i = 0; i < sizeIntCodeSymbols; ++i)
	{
		tmp_binaryCode = inBinary(intCodeSymbols[i]);

		size_t sizeTmpBinaryCode = tmp_binaryCode.size();
		size_t amountAddNull = 24 - sizeTmpBinaryCode;

		for (int i = 0; i < amountAddNull; ++i)
		{
			binaryCodeSymbols.push_back(0);
		}

		for (size_t i = 0; i < sizeTmpBinaryCode; ++i)
		{
			binaryCodeSymbols.push_back(tmp_binaryCode[i]);
		}

		tmp_binaryCode.clear();
	}

	return binaryCodeSymbols;
}


/*******************************************************************************/
int Encryption::fromBinary(vector<int> inBinary)
{
	vector<int> tmp_res;

	size_t sizeInData = inBinary.size();
	size_t degree = sizeInData;
	int res = int();

	for (int i = 0; i < sizeInData; ++i)
	{
		tmp_res.push_back(inBinary[i] * ((int)(pow(2, --degree))));
	}

	for (int i = 0; i < sizeInData; ++i)
	{
		res += tmp_res[i];
	}

	return res;
}


/*******************************************************************************/
vector<int> Encryption::strBinaryToIntBinary(string inBinary)
{
	string tmp = string();

	size_t lenInBinary = inBinary.size();

	vector<int> res;

	for (unsigned int i = 0; i < lenInBinary; ++i)
	{
		tmp = inBinary[i];

		int tmp_int_bin = stoi(tmp);

		res.push_back(tmp_int_bin);
	}

	return res;
}


/*******************************************************************************/
vector<char> Encryption::IntbinaryToChar(vector<int> inBinary)
{
	size_t sizeInBinary = inBinary.size();

	int sizeBlock = 24;
	int beginBlock = 0;
	int endBlock = sizeBlock;
	int countBlock = (int)sizeInBinary / sizeBlock;

	vector<int> symbolCodeBinary;
	vector<char> res;

	for (int i = 0; i < sizeInBinary; ++i)
	{
		if (countBlock)
		{
			for (int d = beginBlock; d < endBlock; ++d)
			{
				symbolCodeBinary.push_back(inBinary[d]);
			}

			beginBlock += sizeBlock;
			endBlock += sizeBlock;
			--countBlock;

			res.push_back(fromBinary(symbolCodeBinary));

			symbolCodeBinary.clear();
		}
		else
		{
			break;
		}
	}

	return res;
}


/*******************************************************************************/
vector<int> Encryption::arrayKeys(vector<int> inKey, int amountRounds)
{
	vector<int> rightAraayKeys;
	vector<int> tmpRounderKeys;
	vector<int> readyRounderRey;

	// Number of key expansion
	const int sincerelyNumber = 2;

	size_t sizeInKey = inKey.size();
	size_t halfSizeInKey = sizeInKey / 2;

	// Take the right half of the key
	for (size_t i = halfSizeInKey; i < sizeInKey; ++i)
	{
		rightAraayKeys.push_back(inKey[i]);
	}

	// Calculate the key for each round
	int rounderKey = int();

	for (int i = 1; i <= amountRounds; ++i)
	{
		rounderKey = (int)pow(sincerelyNumber, i);

		tmpRounderKeys = inBinary(rounderKey);

		size_t sizeTmpRounderKey = tmpRounderKeys.size();
		size_t amountBitsToNull = halfSizeInKey - sizeTmpRounderKey;
		size_t sizeReadyKey = sizeTmpRounderKey + amountBitsToNull;

		for (int i = 0; i < amountBitsToNull; ++i)
		{
			readyRounderRey.push_back(0);
		}

		for (int i = 0; i < sizeTmpRounderKey; ++i)
		{
			readyRounderRey.push_back(tmpRounderKeys[i]);
		}

		size_t tmp_sizeRightAraayKeys = rightAraayKeys.size();
		size_t lastRecordedKeyBegin = tmp_sizeRightAraayKeys - sizeReadyKey;

		for (int i = 0; i < sizeReadyKey; ++i)
		{
			readyRounderRey[i] = readyRounderRey[i] ^ rightAraayKeys[lastRecordedKeyBegin];

			++lastRecordedKeyBegin;
		}

		for (int i = 0; i < sizeReadyKey; ++i)
		{
			rightAraayKeys.push_back(readyRounderRey[i]);
		}

		readyRounderRey.clear();
	}

	return rightAraayKeys;
}


/*******************************************************************************/
void Encryption::subblock(vector<int> inBlock, vector<int>& leftBlock, vector<int>& rightBlock)
{
	size_t sizeInBlock = inBlock.size();
	size_t halfSizeInBlock = sizeInBlock / 2;

	// Add values in left block 
	for (int i = 0; i < halfSizeInBlock; ++i)
	{
		leftBlock.push_back(inBlock[i]);
	}

	// Add values in right block
	for (size_t i = halfSizeInBlock; i < sizeInBlock; ++i)
	{
		rightBlock.push_back(inBlock[i]);
	}
}


/*******************************************************************************/
vector<int> Encryption::F(vector<int> subblock, vector<int> rightAraayKeys, int numberRounds)
{
	vector<int> res;

	size_t sizeSubblock = subblock.size();
	size_t endKey = size_t();
	size_t beginKey = size_t();

	if (numberRounds == 0)
	{
		endKey = sizeSubblock;
		beginKey = 0;
	}
	else
	{
		endKey = sizeSubblock * (numberRounds + 1);
		beginKey = endKey - sizeSubblock;
	}

	int i = 0;

	for (beginKey; beginKey < endKey; ++beginKey)
	{
		res.push_back(subblock[i] ^ rightAraayKeys[beginKey]);

		++i;
	}

	return res;
}


/*******************************************************************************/
vector<int> Encryption::encryptFeistelNetwork(vector<int> leftBlock, vector<int> rightBlock, const vector<int>& rightAraayKeys, int amountRounds)
{
	vector<int> res;
	vector<int> temp;
	vector<int> tempResF;

	size_t sizeRightBlock = rightBlock.size();

	for (int i = 0; i < amountRounds; ++i)
	{
		tempResF = F(leftBlock, rightAraayKeys, i);

		for (int k = 0; k < sizeRightBlock; ++k)
		{
			temp.push_back(rightBlock[k] ^ tempResF[k]);
		}

		rightBlock = leftBlock;

		leftBlock = temp;

		temp.clear();
	}

	for (int i = 0; i < sizeRightBlock; ++i)
	{
		res.push_back(leftBlock[i]);
	}

	for (int i = 0; i < sizeRightBlock; ++i)
	{
		res.push_back(rightBlock[i]);
	}

	return res;
}


/*******************************************************************************/
vector<int> Encryption::decryptFeistelNetwork(const vector<int>& inBlock, int amountRounds, const vector<int>& rightAraayKeys)
{
	vector<int> res;
	vector<int> leftBlock;
	vector<int> rightBlock;
	vector<int> temp;
	vector<int> tempResF;

	subblock(inBlock, leftBlock, rightBlock);

	size_t sizeRightBlock = rightBlock.size();

	for (int i = amountRounds - 1; i >= 0; --i)
	{
		tempResF = F(rightBlock, rightAraayKeys, i);

		for (int k = 0; k < sizeRightBlock; ++k)
		{
			temp.push_back(leftBlock[k] ^ tempResF[k]);
		}

		leftBlock = rightBlock;

		rightBlock = temp;

		temp.clear();
	}

	for (int i = 0; i < sizeRightBlock; ++i)
	{
		res.push_back(leftBlock[i]);
	}

	for (int i = 0; i < sizeRightBlock; ++i)
	{
		res.push_back(rightBlock[i]);
	}

	return res;
}